docs/model.rst

```diff
@@ -31,34 +162,317 @@ persistence driver. Use it to define fields of your model::
+
 .. php:method:: addField($name, $defaults)
 
-    Creates a new field objects inside your model (by default the class is 'Field').
-    The fields are implemented on top of Containers from Agile Core.
+Creates a new field objects inside your model (by default the class is 'Field').
+The fields are implemented on top of Containers from Agile Core.
 
-    Second argument to addField() can supply field default properties::
+Second argument to addField() will contain a seed for the Field class::
 
-        $this->addField('surname', ['default'=>'Smith']);
+   $this->addField('surname', ['default'=>'Smith']);
 
-.. php:method:: addCalculatedField($name, $callback)
+Additionally, `type` property can be used to determine the best `Field` class to handle
+the type::
+
+   $field = $this->addField('is_married', ['type'=>'boolean']);
+
+   // $field type is Field\Boolean
 
-    Creates new field object inside your model. Field value will be automatically
-    calculated by your callback method right after loading data record in model::
+You may also specify your own Field implementation::
 
-        $this->addCalculatedField('total', function($m) {
-            return $m['net'] + $m['vat'];
-        });
+   $field = $this->addField('amount_and_currency', new MyAmountCurrencyField());
 
 Read more about :php:class:`Field`
 
+Read-only Fields
+^^^^^^^^^^^^^^^^
+Although you may make any field read-only::
+
+   $this->addField('name', ['read_only'=>true]);
+
+There are two methods for adding dynamically calculated fields.
+
+.. php:method:: addExpression($name, $definition)
+
+Defines a field as server-side expression (e.g. SQL)::
+
+   $this->addExpression('total', '[amount] + [vat]');
+
+The above code is executed on the server (SQL) and can be very powerful.
+You must make sure that expression is valid for current `$this->persistence`::
+
+   $product->addExpression('discount', $this->refLink('category_id')->fieldQuery('default_discount'));
+   // expression as a sub-select from referenced model (Category) imported as a read-only field
+   // of $product model
+
+   $product->addExpression('total', 'if([is_discounted], ([amount]+[vat])*[discount], [amount] + [vat])');
+   // new "total" field now contains complex logic, which is executed in SQL
+
+   $product->addCondition('total', '<', 10);
+   // filter products that cost less than 10.00 (including discount)
+
+
+For the times when you are not working with SQL persistence, you can calculate field in PHP.
+
+.. php:method:: addCalculatedField($name, $callback)
+
+Creates new field object inside your model. Field value will be automatically
+calculated by your callback method right after loading data record in model::
+
+   $this->addField('term', ['caption'=>'Repayment term in months', 'default'=>36]);
+   $this->addField('rate', ['caption'=>'APR %', 'default'=>5]);
+
+   $this->addCalculatedField('interest', function($m) use($apr_calculator) {
```

Comment:
To make the sample code consise, I think `$apr_calculator` should be defined in sample code.
Or make code easier without `use()` and just have `return $m->calculateInterest();` ?